spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -384,6 +408,24 @@ else if (result instanceof Message) {
 				if (sourceIsMessage) {
 					MessageBuilder<Object> builder = MessageBuilder.withPayload(result)
 							.setHeader(KafkaHeaders.TOPIC, topic);
+					if (this.replyHeadersConfigurer != null) {
+						Map<String, Object> headersToCopy = ((Message<?>) source).getHeaders().entrySet().stream()
+							.filter(e -> {
+								String key = e.getKey();
+								return !key.equals(MessageHeaders.ID) && !key.equals(MessageHeaders.TIMESTAMP)
+										&& !key.equals(KafkaHeaders.CORRELATION_ID)
+										&& !key.startsWith(KafkaHeaders.RECEIVED);
+							})
+							.filter(e -> this.replyHeadersConfigurer.shouldCopy(e.getKey(), e.getValue()))
+							.collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue()));
+						if (headersToCopy.size() > 0) {
+							builder.copyHeaders(headersToCopy);
+						}
+						headersToCopy = this.replyHeadersConfigurer.additionalHeaders();
+						if (!ObjectUtils.isEmpty(headersToCopy)) {
+							builder.copyHeaders(headersToCopy);
```

Comment:
maybe `copyHeadersIfAbsent()` ?

Reply:
I don't think so; this is a new `MessageBuilder` so there won't be any existing headers - except for those that passed the filter above (plus the correlation id and topic).

What concerns you?

Reply:
Exactly! That’s my point that end-user May shot his/her own leg when the same header passed the filter and it is in the extra. Maybe just add a JavaDoc note that these additional always win?..